### PR TITLE
Lighthouse a11y updates

### LINKED ISSUE #1629: content: New a11y audit docs for Lighthouse 6.0
**Author**: @mfriesenhahn 
**Reviewer**: @kaycebasques 

[List of docs](https://github.com/GoogleChrome/lighthouse/issues/7127#issue-405436641)

**Target publish date:** 2019-10-17 (Chrome 79 Branch Point)

- [x] Check this box if this is a hard deadline.
- [ ] Check this box if you'd like your reviewer to fix grammar and punctuation issues for you.

**Process**

Check out the [web.dev writing process](https://web.dev/handbook/quick-start/).

When you're ready, make a copy of [this template](https://docs.google.com/document/d/1lgaNIEnXZf-RB8_p9RK22QEgpXJqnu77pLWVWVy4nuw/edit)
and fill in the proposal and outline sections. When you're finished, drop a link
to the doc in the 1st step below and check the box. A member of the web.dev team
will assign this ticket to themselves and work as your reviewer to help you
complete the remaining steps.

- [x] 1. I've done my proposal <add a link to your proposal doc>
- [x] 2. Proposal approved — _reviewer checks this_
- [x] 3. I've added a draft to my proposal doc
- [x] 4. Draft approved — _reviewer checks this_
- [x] 5. I've submitted a pull request #1648
- [x] 6. Pull request approved — _reviewer checks this_

### PR DESCRIPTION
**Pull request by mfriesenhahn**
Fixes #1629 and #1341.

Changes proposed in this pull request:

- Add Lighthouse 6.0 a11y audit posts (#1629).
- Normalize and improve content quality of existing Lighthouse a11y posts (#1341).
- Deprecate `heading-levels` post. (Depends on PR #1646.)
- Update table in `labels-and-text-alternatives` to refer to element types rather than LH audits to avoid potential confusion and future maintenance errors.
- Add various includes in `src/site/includes/content/lighthouse-accessibility` to DRY up a11y audit posts, which have significant overlap due to granularity of axe checks.
- Minor updates to `template-audit-post`.

### CODE REVIEW

**Comment by mfriesenhahn** [comment]
@kaycebasques, I've submitted this so you can start reviewing the new a11y audits, but the older posts aren't ready for review yet. I'll update the status of #1341 once I'm done with the updates.

Status notes about new audits to help you gauge when to review:
- @robdodson and the LH team are still evaluating the behavior of the axe `heading-order` check, so it's not yet certain that it'll make the cut. However, the content is so similar to the preexisting `heading-levels` manual audit that I just went ahead and cleaned it up. If it isn't used, I think we can pretty much just migrate the content over to the old audit.
- After some discussion Rob suggested (and I agree) that `html-xml-lang-mismatch` be cut since it covers a very unusual case. I had already written the post at that point, so I've left it in just in case.
- I've [suggested](https://github.com/GoogleChrome/lighthouse/pull/9798#issuecomment-540294032) that `duplicate-id`, `duplicate-id-active`, and `duplicate-id-aria` be consolidated into a single LH audit, which I think will be easier for people to understand. Still waiting to hear back from the LH team about that.

**Comment by mfriesenhahn** [comment]
Update:
- `heading-order` will replace `heading-levels` and will take the failed state of its title: "Headings skip levels"
- I've removed `html-xml-lang-mismatch` from the PR.
- `duplicate-id` has been deprecated. `duplicate-id-aria` and `duplicate-id-active` will be kept as separate audits.

cc @kaycebasques, @robdodson, @brendankenny

File: src/site/content/en/lighthouse-accessibility/list/index.md [hunk 1]
```diff
@@ -1,43 +1,41 @@
...
-remove any elements in lists that don't belong there.
+## How to create correctly structured lists
+
+Remove any elements in lists that don't belong there.
 Ordered and unordered lists must only contain `<li>`, `<script>` or `<template>` elements.
 
-Valid lists must have parent elements (`ul` or `ol` elements) and child elements (`li` elements).
+Valid lists must have parent elements (`<ul>` or `<ol>` elements) and child elements (`<li>` elements).
 Any other content elements are invalid.
 Learn more in
```
Comment (robdodson) [inline src/site/content/en/lighthouse-accessibility/list/index.md]: Just noticed this, but I think many audits in this section contain a link to the deque doc, and then another link to it in the Resources section immediately after.

File: src/site/content/en/lighthouse-accessibility/layout-table/index.md [hunk 2]
```diff
@@ -1,32 +1,37 @@
...
+
+{% include 'content/lighthouse-accessibility/scoring.njk' %}
+
+## How to make presentational tables accessible
 
 It's simple to pass the Lighthouse audit:
-remove data elements (`th`, `caption` elements or the `summary` attribute) from layout tables.
+remove data elements (`<th>`, `<caption>` elements or the `summary` attribute) from layout tables.
 
 If using a table just for visual layout, and not for tabular data,
```
Comment (robdodson) [inline src/site/content/en/lighthouse-accessibility/layout-table/index.md]: Maybe we should replace this section by telling folks to use CSS grid or Flexbox.

File: src/site/content/en/lighthouse-accessibility/td-headers-attr/index.md [hunk 3]
```diff
@@ -1,9 +1,11 @@
 ---
 layout: post
-title: Ensure there's only one table header per table column
+title: "Cells in a `<table>` element that use the `[headers]` attribute refer to an element id not found within the same table"
```
Comment (robdodson) [inline src/site/content/en/lighthouse-accessibility/td-headers-attr/index.md]:
```suggestion
title: "Cells in a `<table>` element that use the `[headers]` attribute refer to an element ID not found within the same table"
```

**Comment by robdodson** [comment]
Some recurring comments:

- We probably want to find usage of "Screen reader users" and replace with "Users of assistive technology.
- I noticed a lot of places where it says "Learn more <link to some deque doc>" and then immediately in the Resources section we show the same link. Do we want to consolidate these into just the Resources section?

File: src/site/content/en/lighthouse-accessibility/focus-traps/index.md [hunk 4]
```diff
@@ -1,9 +1,11 @@
 ---
 layout: post
```
Comment (robdodson) [inline src/site/content/en/lighthouse-accessibility/focus-traps/index.md]: This post starts by mentioning the web aim checklist. But that might seem a bit odd because I don't think we mention it elsewhere.

File: src/site/content/en/lighthouse-accessibility/custom-control-roles/index.md [hunk 5]
```diff
@@ -1,11 +1,13 @@
 ---
 layout: post
```
Comment (robdodson) [inline src/site/content/en/lighthouse-accessibility/custom-control-roles/index.md]: Wanted to mention that the Glitch example in this post is currently not working. I need to fix it.

File: src/site/content/en/lighthouse-accessibility/offscreen-content-hidden/index.md [hunk 6]
```diff
@@ -1,9 +1,11 @@
 ---
```
Comment (robdodson) [inline src/site/content/en/lighthouse-accessibility/offscreen-content-hidden/index.md]: Can't comment on it because it's not part of this PR but I noticed this doc opens with:

"For instance, content that's offscreen or just presentation should be hidden from a screen reader."

I think it should be: "For instance, content that's offscreen or just **presentational** should be hidden from a screen reader."

File: src/site/content/en/lighthouse-accessibility/use-landmarks/index.md [hunk 7]
```diff
@@ -13,7 +15,7 @@ HTML5 elements such as `main`, `nav`, and `aside` act as landmarks,
 or special regions on the page to which a screen reader can jump.
 By using landmark elements,
 you can dramatically improve the navigation experience on your site
-for users of assitive technology.
+for users of assistive technology.
 Learn more in
 [HTML 5 and ARIA Landmarks](https://dequeuniversity.com/assets/html/jquery-summit/html5/slides/landmarks.html).
```
Comment (robdodson) [inline src/site/content/en/lighthouse-accessibility/use-landmarks/index.md]: What do you think about the example below? It's using `role` (I'm assuming as a fallback for missing landmark roles) but I'm not sure if it actually needs to do that. I'm pretty sure modern browsers support the landmark elements in that example.

File: src/site/content/en/lighthouse-accessibility/button-name/index.md [hunk 8]
```diff
@@ -1,56 +1,54 @@
...
+  Learn how to improve the accessibility of your web page by making sure that
+  all buttons have names that screen readers can access.
 date: 2019-05-02
+updated: 2019-09-19
 web_lighthouse:
   - button-name
 ---
 
 When a button doesn't have an accessible name,
 screen readers announce it as "button",
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/button-name/index.md]: Yeah, I haven't done a quality pass on the "Names and Labels" section yet.

Here's what I did:

"When a button doesn't have an accessible name, screen readers and other assistive technologies announce it as _button_, which provides no information to users about what the button does."

File: src/site/content/en/lighthouse-accessibility/button-name/index.md [hunk 9]
```diff
@@ -1,56 +1,54 @@
...
+<button>Book room</button>
 ```
 
-When working with icon buttons,
+For buttons without visible labels, like icon buttons,
 use the `aria-label` attribute to clearly describe the action
 to anyone using a screen reader, for example:
 
 ```html
 <button aria-label="Bold text icon"></button>
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/button-name/index.md]: Agree. Might even make sense to do this as a Glitch with the CSS so folks can get a full sense of what this is?

File: src/site/content/en/lighthouse-accessibility/color-contrast/index.md [hunk 10]
```diff
@@ -1,47 +1,100 @@
...
+Text that has a low contrast ratio—that is,
+text whose brightness is too close to the background brightness—can be
+hard to read.
+For example, presenting light gray text on a white background
+makes it difficult for users to distinguish the shapes of the characters,
+which can reduce reading comprehension and slow down reading speed.
+
+While this issue is particularly challenging for people with low vision,
+low-contrast text can negatively affect the reading experience
+for all your users.
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/color-contrast/index.md]: Nope, you use _all of_ prior to a pronoun and _all_ prior to a noun. See https://dictionary.cambridge.org/grammar/british-grammar/all

Though _all of [noun]_ is also considered grammatical these days--I'm guessing it's one of those cases where over-generalization of a narrow grammar rule entered common usage and became accepted.

File: src/site/content/en/lighthouse-accessibility/color-contrast/index.md [hunk 11]
```diff
@@ -1,47 +1,100 @@
...
-Try the Color Contrast Analyser in
-[Text elements must have sufficient color contrast against the background](https://dequeuniversity.com/rules/axe/3.3/color-contrast).
+Once you have a passing color value, update your project's CSS.
 
-<!--
-## How this audit impacts overall Lighthouse score
+More complex cases, like text on a gradient or text on an image,
+need to be checked manually,
+as do UI elements and images.
+Consider using a tool like the Paciello Group's
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/color-contrast/index.md]: True for HTML text on an image, but I was wanting to provide a solution for raster text and other images and was worried about length/complexity. To use the color picker eyedropper in the latter cases requires you to first inspect some HTML text and then eye drop the image you're actually interested in, which is pretty cumbersome.

I could break out the HTML-text-on-an-image case, but I kinda think presenting one tool that you can use for multiple cases is more straightforward. But happy to do whatever you think best here!

File: src/site/content/en/lighthouse-accessibility/list/index.md (hunk 1 above)
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/list/index.md]: Yup. As noted on another comment, that's something I was cleaning up as I did a quality pass on each section. I can go ahead and remove those globally.

File: src/site/content/en/lighthouse-accessibility/accesskeys/index.md [hunk 12]
```diff
@@ -1,50 +1,47 @@
...
 
-Todo. I have no idea how accessibility scoring is working!
--->
-## More information
+## Resources
 
-- [Access key values are unique audit source](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/accessibility/accesskeys.js)
-- [axe-core rule descriptions](https://github.com/dequelabs/axe-core/blob/develop/doc/rule-descriptions.md)
+- [Source code for **`[accesskey]` values are not unique** audit](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/accessibility/accesskeys.js)
+- [accesskey attribute value must be unique](https://dequeuniversity.com/rules/axe/3.3/accesskeys)
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/accesskeys/index.md]: Fixed in all posts by adding _(Deque University)_ to the end of the link text.

File: src/site/content/en/lighthouse-accessibility/form-field-multiple-labels/index.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+web_lighthouse:
+  - form-field-multiple-labels
+date: 2019-10-17
+---
+
+{% include 'content/lighthouse-accessibility/accessible-names.njk' %}
+
+`<label>` elements are a common way to assign names to native controls.
+If you unintentionally associate multiple labels with a single control,
+you'll create inconsistent behavior across browsers and screen readers:
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/form-field-multiple-labels/index.md]: So, in this case, axe specifically refers to browser-screen reader combos. I don't know enough about other assistive tech to know whether they have the same inconsistent behaviors. Can you confirm?

File: src/site/content/en/lighthouse-accessibility/frame-title/index.md [hunk 14]
```diff
@@ -46,13 +47,8 @@ Learn more in [Frames must have title attribute](https://dequeuniversity.com/rul
 Learn more in
 [Write descriptive titles, descriptions, and link text for every page](/write-descriptive-text).
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/frame-title/index.md]: I'd leave it here, FWIW, since it seems directly related to the section ("Tips for creating descriptive frame titles").

File: src/site/content/en/lighthouse-accessibility/input-image-alt/index.md [hunk 15]
```diff
@@ -34,8 +36,7 @@ in the `alt` text
     Username:
     <input type="text">
   </label>
-  <input type="image" alt="Sign in"
-             src="./sign-in-button.png">
+  <input type="image" alt="Sign in" src="./sign-in-button.png">
 </form>
 ```
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/input-image-alt/index.md]: Yeah, it makes it sound like it's _less_ important to provide alt text in this case rather than more. Plus, if an image includes _only_ text, it really shouldn't be an image. Cut.

File: src/site/content/en/lighthouse-accessibility/color-contrast/index.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (robdodson) [reply src/site/content/en/lighthouse-accessibility/color-contrast/index.md]: I didn't know their tool handled raster text, or essentially image over image. That's neat.
Personally I think it's useful to give a shout out to the eyedropper because positioning text over an image with CSS is pretty common, and having to install a native app to check that contrast feels a bit burdensome.

Kind of a bummer that we don't mention that functionality here: https://developers.google.com/web/tools/chrome-devtools/accessibility/reference#contrast
I know Alice spent a decent amount of time building it, which is why I wanted to highlight it.

Is it possible to add an Aside note to mention that you can use the eyedropper in devtools, or the Pacielllo app, to test against image backgrounds?

File: src/site/content/en/lighthouse-accessibility/form-field-multiple-labels/index.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (robdodson) [reply src/site/content/en/lighthouse-accessibility/form-field-multiple-labels/index.md]: I don't know on this specific issue. I do know that things like Dragon voice assistant do rely on the accessibility tree and labels when you say things like "Click on menu button". If the button isn't properly labeled "menu", it won't work. My hunch is this issue would affect voice assistants as well...

**Comment by robdodson** [comment]
@mfriesenhahn feel free to ping me on chat for any outstanding questions I may have missed.

File: src/site/content/en/lighthouse-accessibility/duplicate-id-aria/index.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    alt="Lighthouse audit showing ARIA elements with duplicate IDs">
+</figure>
+
+This audit is similar to the
+[**`[id]` attributes on active, focusable elements are not unique**](/duplicate-id-active) audit
+but checks for duplicate IDs in a different set of elements.
+
+{% include 'content/lighthouse-accessibility/scoring.njk' %}
+
+{% include 'content/lighthouse-accessibility/fix-duplicate-ids.njk' %}
```
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/duplicate-id-aria/index.md]: Yeah, was moving too quickly. Good catch.

File: src/site/content/en/lighthouse-accessibility/layout-table/index.md (hunk 2 above)
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/layout-table/index.md]: I'm going to leave this for another PR where I take a pass at the tables and lists section as a whole.

File: src/site/content/en/lighthouse-accessibility/focus-traps/index.md (hunk 4 above)
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/focus-traps/index.md]: Agree. Don't see any reason why we need to cite our source in this one particular case and no others.

File: src/site/content/en/lighthouse-accessibility/use-landmarks/index.md (hunk 7 above)
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/use-landmarks/index.md]: Yeah, that's weird. Removing the roles.

File: src/site/content/en/lighthouse-accessibility/button-name/index.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/button-name/index.md]: @robdodson, like this, maybe: https://glitch.com/edit/#!/lh-button-name-demo

The Material font uses ligature magic, which may not be obvious to everyone. Not sure whether that should be called out--or maybe use SVG instead?

**Comment by robdodson** [comment]
Going to merge this so we can stage the links for LH folks.

File: src/site/content/en/lighthouse-accessibility/button-name/index.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (robdodson) [reply src/site/content/en/lighthouse-accessibility/button-name/index.md]: I think that's a cool example to include. Using the font is ok if we add a comment. I think that's a common scenario that illustrates when this technique would be useful.

File: src/site/content/en/lighthouse-accessibility/color-contrast/index.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (mfriesenhahn) [reply src/site/content/en/lighthouse-accessibility/color-contrast/index.md]: > I didn't know their tool handled raster text, or essentially image over image. That's neat.

I mean, it just grabs the color from any pixel you point it at.

> Personally I think it's useful to give a shout out to the eyedropper because positioning text over an image with CSS is pretty common, and having to install a native app to check that contrast feels a bit burdensome.

I take your point, though I personally find it less burdensome to install a small native app that can check everything than to use multiple tools for slightly different cases. That said, I'll put in a note about the background color picker!
